components/metrics/lib.rs

```diff
@@ -1,77 +1,278 @@
+    all_interactive: bool,
+}
+
+
+//TODO i don't think all interactive is needed anymore
+impl InteractiveWindow {
+    pub fn new() -> InteractiveWindow {
+        InteractiveWindow {
+            start: precise_time_ns(),
+            instant: Instant::now(),
+            all_interactive: false,
+        }
+    }
+
+    // We need to either start or restart the 10s window
+    //   start: we've added a new document
+    //   restart: there was a task > 50s
+    //   not all documents are interactive
+    pub fn start_window(&mut self) {
+        self.start = precise_time_ns();
+        self.instant = Instant::now();
+        self.unset_all_interactive();
+    }
+
+    pub fn needs_check(&self) -> bool {
+        self.instant.elapsed() > *TEN_SECONDS
+    }
+
+    pub fn check_interactive(&self) -> bool {
+        self.all_interactive
+    }
+
+    pub fn set_all_interactive(&mut self) {
+        self.all_interactive = true;
+    }
+
+    pub fn unset_all_interactive(&mut self) {
+        self.all_interactive = false;
+    }
+
+    pub fn get_start(&self) -> u64 {
+        self.start
+    }
+
+    pub fn elapsed_since_start(&self) -> u64 {
+        self.instant.elapsed().as_secs()
+    }
+
+    // ^^^^^^^^^^^^^ expected u64, found struct `MAX_TASK_TIME`
+
+    pub fn max_task_time() -> u64 {
+        50000000
+    }
+}
+
+#[derive(Debug)]
+pub enum InteractiveFlag {
+    DCL,
+    TTI,
+}
+
+impl InteractiveMetrics {
+    pub fn new(time_profiler_chan: ProfilerChan) -> InteractiveMetrics {
+        InteractiveMetrics {
+            navigation_start: Cell::new(None),
+            dom_content_loaded: Cell::new(None),
+            main_thread_available: Cell::new(None),
+            time_to_interactive: Cell::new(None),
+            time_profiler_chan: time_profiler_chan,
+        }
+    }
 
-            let navigation_start = match self.navigation_start {
-                Some(time) => time,
-                None => {
-                    warn!("Trying to set metric before navigation start");
-                    return;
+    pub fn set_dom_content_loaded(&self) {
+        if self.dom_content_loaded.get().is_none() {
+            self.dom_content_loaded.set(Some(precise_time_ns() as f64));
+        }
+    }
+
+    pub fn set_main_thread_available(&self, time: Option<f64>) {
+        if self.main_thread_available.get().is_none() && time.is_some() {
+            self.main_thread_available.set(time);
+        }
+    }
+
+    pub fn get_dom_content_loaded(&self) -> Option<f64> {
+        self.dom_content_loaded.get()
+    }
+
+    pub fn get_main_thread_available(&self) -> Option<f64> {
+        self.main_thread_available.get()
+    }
+
+    // can set either dlc or tti first, but both must be set to actually calc metric
+    // when the second is set, set_tti is called with appropriate time
+    pub fn maybe_set_tti<T>(
+        &self,
+        profiler_metadata_factory: &T,
+        time: Option<f64>,
+        metric: InteractiveFlag,
```

Comment:
Let's combine the time and metric arguments like this:
```rust
enum InteractiveFlag {
    DomContentLoaded,
    TimeToInteractive(f64),
}
```
This will make some of the related code clearer.